man/man1/p11sak.1.in

```diff
@@ -426,6 +446,27 @@ to use a configuration for the private key only, the string has to start with th
 .
 .
 .
+.SS "\-\-custom\-attr | \-c [[pub_attrs]:[prv_attrs]]"
+sets the binary attributes of the key.
+.PP
+.B Note:
+not all binary attributes are applicable to all keys and will be omitted if not applicable. 
+.PP
+.B Syntax:
+For multiple attributes, seperate them with a ",", the id can be given in decimal or hexadecimal
+and the valid values are "true" or "false": e. g. '368=true,0x00000163=false,0x00000105=false'.
+From Example above: CKA_MODIFIABLE=true, CKA_LOCAL=false, CKA_DECRYPT=true
```

Comment:
Wouldn't it make sense that besides specifying the attribute ID in hex or decimal, that I can also specify 'CKA_...' for those that are known? You can use the attribute definitions from the config file for this (and check if they are of type BOOL, too).

Reply:
also for ULONG attributes of course....

Reply:
I still think we should generate a lookup structure for attributes similar to the one we have for mechanisms.  Would solve this damn "what is the correct number" problem...

Reply:
I like the idea of having such an attribute lookup table. It could contain also the data type of the attribute, and maybe to which object types it is applicable, etc.

However: Adding such a lookup table makes the just introduced p11sak-config file obsolete. The definitions in the config file are then duplicate to the info from the table (and I guess the info from the table would be more trustful/correct than those from the config file anyway).

The only real use of the config file would then be to let the user define which attributes he/she/it wants to see in the long list form. But actually, why not simply list all attributes that the object has....
So this then makes the config file totally superfluous.

I actually would like to get rid of the config file, because one can only add attributes to it that OCK 'knows'. Adding attributes that OCK does not know will in the best case simply cause it to never actually list that attribute.
Having the lookup table means we as-of-a-sudden support all attributes that OCK knows (maybe p11sak only lists those with certain data types for now), without the user needing to define anything.

The only question is how can we keep backward compatible with the current behavior? Simply ignore the config file?

Opinions? @jschmidb @juergenchrist

Reply:
> why not simply list all attributes that the object has
Yes, indeed. This would also make the newly introduced rightmost column in the short list output superfluous. 
What would we loose? At the moment, the config file is also something like a filter. It defines which additional attributes to list. Removing the config file, we would loose this filter function and we would simply list all attributes of the key object.
So we might keep the config file, but only to list the attributes to be displayed in the long format (well, this could be used to keep the output small).
The config file entries could then be simpler, because we know the hex code and data type from the lookup table. So the config file would just contain a list of attributes (or different lists for selected key types?)

Reply:
There is maybe one good reason to keep the config file approach as it is currently: 

Although the p11sak tool comes as part of OCK, it until now should be usable with any PKCS#11 implementation, not only with OCK.  It uses environment variable PKCSLIB to know which PKCS#11 library to load (the default is libopencryptoki.so). 

Once we introduce the attribute lookup table, and only use that table to know about the attributes ids, data types, etc, p11sak will then no longer be able to display custom/vendor attributes of another PKCS#11 implementation that OCK does not know. 

So we might certainly use the lookup table for the standard attributes (those are defined by the PKCS#11 standard, and thus are common across all PKCS#11 implementation), but for the vendor specific attributes (where the high order bit is on to indicate a vendor attribute) we should at least prefer a definition from the config file over the lookup table, so that users can define their own attributes independent from what OCK knows.